### PR TITLE
Updated SongAPI Closes #83

### LINKED ISSUE #83: Update song vector according to user feedback
Using the feedback supplied by thumbs up/down and SAM's to alter the vector accordingly.

### PR DESCRIPTION
**Pull request by JeremybellEU**
Updated SongAPI according to https://github.com/MoodCat/MoodCat.me/issues/47

- Created classification method in order to let users classify a song
- Created vote method in order to let users vote on songs
- Tested both methods in the SongAPITest
- Modified VAVector constructor to assert that valence and arousal are in the correct range

### CODE REVIEW

File: src/main/java/me/moodcat/api/SongAPI.java [hunk 1]
```diff
@@ -48,4 +62,100 @@ public Song getSongById(@PathParam("id") final int id) {
...
+    @Transactional
+    public ClassificationRequest classifySong(@PathParam("id") final int id,
+            final ClassificationRequest classification) {
+        final Song song = this.songDAO.findById(id);
+
+        if (song.getNumberOfPositiveVotes() < MINIMUM_NUMBER_OF_POSITIVE_VOTES) {
+            final VAVector classificationVector = new VAVector(
+
+                    classification.getValence(),
+                    classification.getArousal());
```
Comment (Jaapp-) [inline src/main/java/me/moodcat/api/SongAPI.java]: It's probably a good idea to add a check whether the valence and arousal have valid values here, and ignore the vector if it's not valid. Else either the asserts of the VAVector will fail (and throw an exception), or any non-valid values will be used.

File: src/main/java/me/moodcat/database/controllers/SongDAO.java [hunk 2]
```diff
@@ -44,6 +44,13 @@ public Song findById(final int id) {
                 .where(song.id.eq(id))
                 .singleResult(song);
     }
+    
+    @Transactional
```
Comment (newnottakenname) [inline src/main/java/me/moodcat/database/controllers/SongDAO.java]: @JeremybellEU Toch mis je hier nog javadoc

File: src/main/java/me/moodcat/api/SongAPI.java [hunk 3]
```diff
@@ -48,4 +62,100 @@ public Song getSongById(@PathParam("id") final int id) {
...
+     * @return The classification provided.
+     */
+    @POST
+    @Path("{id}/classify")
+    @Transactional
+    public ClassificationRequest classifySong(@PathParam("id") final int id,
+            final ClassificationRequest classification) {
+        final Song song = this.songDAO.findById(id);
+
+        if (song.getNumberOfPositiveVotes() < MINIMUM_NUMBER_OF_POSITIVE_VOTES) {
```
Comment (Azkle) [inline src/main/java/me/moodcat/api/SongAPI.java]: Shouldn't this be a separate method, I mean to calculate the new arousal/valence isn't something that should be in a POST method.

File: src/main/java/me/moodcat/api/SongAPI.java [hunk 4]
```diff
@@ -48,4 +71,131 @@ public Song getSongById(@PathParam("id") final int id) {
...
+        @Getter
+        private final double arousal;
+    }
+
+    /**
+     * Thrown if the classification was invalid.
+     *
+     * @author JeremybellEU
+     */
+    protected static class InvalidClassificationException extends Exception {
```
Comment (JWGmeligMeyling) [inline src/main/java/me/moodcat/api/SongAPI.java]: Please either use `IllegalArgumentException` or extend `IllegalArgumentException` for consistency.

File: src/main/java/me/moodcat/api/SongAPI.java [hunk 5]
```diff
@@ -48,4 +71,131 @@ public Song getSongById(@PathParam("id") final int id) {
...
+         */
+        private static final long serialVersionUID = -6684926632173744801L;
+    }
+
+    /**
+     * Thrown if the vote was invalid.
+     *
+     * @author JeremybellEU
+     */
+    protected static class InvalidVoteException extends Exception {
```
Comment (JWGmeligMeyling) [inline src/main/java/me/moodcat/api/SongAPI.java]: Same here.

File: src/main/java/me/moodcat/database/entities/Song.java [hunk 6]
```diff
@@ -68,4 +69,21 @@
     })
     private VAVector expectedValenceArousal;
 
+    @JsonIgnore
```
Comment (JWGmeligMeyling) [inline src/main/java/me/moodcat/database/entities/Song.java]: Are we still abusing our entities as models?

**Comment by JWGmeligMeyling** [comment]
Looks good and well tested. Please separate fixes & clean up in future prs. Please fix the remarks and then it can be merged!

**Comment by Azkle** [comment]
Good to merge!

**Pull request merged**